### DISCUSSION TITLE
透明代理下sniffing功能提升

### DISCUSSION THREAD

**Discussion by XGFan** [v2fly/v2ray-core#1635] [Ideas]
在我的理解里，sniffing功能应该有以下用途
1，修正dns劫持
2，sniffing出域名，根据域名进行路由（透明代理场景下

现在实现里其实只有1的功能。
https://github.com/v2fly/v2ray-core/blob/65e425ecf5b5a818a4a3e4b40448fdf7b1b8b6b2/app/dispatcher/default.go#L214-L234

只有符合shouldOverride（即开启destOverride）
sniffing出的域名才会参与后续的`d.routedDispatch `

不管是否override dest，只要开启sniffing，都应该让域名参与路由分配。

**Comment by dyhkwong** [comment]
Do you mean something like #1271?

**Reply by XGFan** [reply]
> Do you mean something like #1271?

是的，实现起来差不多，其实并不需要多一个参数。
只需要在outbound上多一个RoutingTarget就好了。处理路由的时候优先通过ob.RoutingTarget来进行就好了
大致就这样
```
if err == nil {
	domain := result.Domain()
	destination.Address = net.ParseAddress(domain)
	if shouldOverride(result, sniffingRequest.OverrideDestinationForProtocol) {
		ob.Target = destination
	} else {
		ob.RoutingTarget = destination
	}
}
```

**Comment by mokeyish** [comment]
我这将 dns 请求转发至 dns_in，dns_in 写了 sniffing 然后路由也配置了  dns_in->dns_out，v2ray 应该是知道 ip 对应的域名是什么的，

但配置域名规则，路由完全没用，全走默认代理了。

**Reply by dyhkwong** [reply]
dns_in 写 sniffing 当然没用，需要嗅探有 http/tls/quic 进入的入站才能取得对应的域名。

**Reply by mokeyish** [reply]
> dns_in 写 sniffing 当然没用，需要嗅探有 http/tls/quic 进入的入站才能取得对应的域名。

我任意门那里，也加了 sniffing 的。

不过 sniffing 不应该是全局的嘛，进入了 dns 路由模块，就应该记下来。。。
